lib/ClangImporter/ImportDecl.cpp

```diff
@@ -3434,12 +3443,19 @@ namespace {
 
       bool hasReferenceableFields = !members.empty();
 
-      if (hasZeroInitializableStorage) {
-        // Add constructors for the struct.
+      const clang::CXXRecordDecl *cxxRecordDecl =
+          dyn_cast<clang::CXXRecordDecl>(decl);
+      if (hasZeroInitializableStorage && !cxxRecordDecl) {
+        // Add default constructor for the struct if compiling in C mode.
+        // If we're compiling for C++, we'll import the C++ default constructor
+        // (if there is one), so we don't need to synthesize one here.
         ctors.push_back(createDefaultConstructor(Impl, result));
       }
 
-      if (hasReferenceableFields && hasMemberwiseInitializer) {
+      bool hasUserDeclaredConstructor =
+          cxxRecordDecl && cxxRecordDecl->hasUserDeclaredConstructor();
+      if (hasReferenceableFields && hasMemberwiseInitializer &&
+          !hasUserDeclaredConstructor) {
```

Comment:
This is creating a memberwise initializer, right?  Or at least a partial one for the members we can import?  For C++, this is strongly analogous to aggregate initialization, and the condition we should be using is `isAggregate()`.  If that's wrong, we at least need to be explaining why in a comment.

Reply:
Makes sense to me. Fixed.

Reply:
Actually, there is a slight difference. If a struct has a deleted constructor, it's still an aggregate, but (with the `hasUserDeclaredConstructor` condition above), we wouldn't generate a "memberwise initializer." C++ [allows aggregate initialization even if there's an explicitly deleted constructor](https://godbolt.org/z/v45sWs), so I think this is the correct behavior. But I figured it would be a good to point it out anyway.